hazelcast/src/main/java/com/hazelcast/scheduledexecutor/impl/ScheduledExecutorContainer.java

```diff
@@ -0,0 +1,332 @@
+        checkNotStaleTask(taskName);
+        ScheduledTaskDescriptor descriptor = tasks.get(taskName);
+        return descriptor.getScheduledFuture().isCancelled();
+    }
+
+    public boolean isDone(String taskName) {
+        checkNotStaleTask(taskName);
+        ScheduledTaskDescriptor descriptor = tasks.get(taskName);
+        return descriptor.getScheduledFuture().isDone();
+    }
+
+    public void destroy(String taskName) {
+        checkNotStaleTask(taskName);
+
+        ScheduledTaskDescriptor descriptor = tasks.get(taskName);
+        ScheduledFuture scheduledFuture = descriptor.getScheduledFuture();
+        if (!scheduledFuture.isDone()) {
+            scheduledFuture.cancel(true);
+        }
+
+        tasks.remove(taskName);
+    }
+
+    public void stash(TaskDefinition definition) {
+        // TODO tkountis - Consider whether the backup task should also be running or not
+        // - provide use cases where either way is wrong
+        if (!backups.containsKey(definition.getName())) {
+            BackupTaskDescriptor descriptor = new BackupTaskDescriptor(definition);
+            backups.put(definition.getName(), descriptor);
+        }
+    }
+
+    public void unstash(String taskName) {
+        backups.remove(taskName);
+    }
+
+    public Collection<ScheduledTaskDescriptor> getTasks() {
+        return tasks.values();
+    }
+
+    public void syncState(String taskName, Map newState) {
+        backups.get(taskName).setMasterState(newState);
+    }
+
+    public int getDurability() {
+        return durability;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    void promoteStash() {
+        Iterator<Map.Entry<String, BackupTaskDescriptor>> backupEntries =
+                backups.entrySet().iterator();
+
+        while (backupEntries.hasNext()) {
+            Map.Entry<String, BackupTaskDescriptor> entry = backupEntries.next();
+            BackupTaskDescriptor descriptor = entry.getValue();
+            doSchedule(descriptor.getDefinition(), descriptor.getMasterState(), descriptor.getMasterStats());
+            backupEntries.remove();
+        }
+
+    }
+
+    Map<String, BackupTaskDescriptor> prepareForReplication() {
+        Map<String, BackupTaskDescriptor> replicas = new HashMap<String, BackupTaskDescriptor>();
+
+        Iterator<Map.Entry<String, ScheduledTaskDescriptor>> taskEntries =
+                tasks.entrySet().iterator();
+
+        while (taskEntries.hasNext()) {
+            Map.Entry<String, ScheduledTaskDescriptor> entry = taskEntries.next();
+            ScheduledTaskDescriptor descriptor = entry.getValue();
+            ScheduledFuture<?> future = descriptor.getScheduledFuture();
+
+            // Stop primary one, so we can safely capture latest state.
+            // However, may be incorrect state due to interruption.
+            while (!future.isCancelled() && !future.isDone()) {
+                future.cancel(true);
```

Comment:
we should not stop the primary one if we cannot resume it afterwards. this replication does not mean we are moving the task, it can happen anytime when a backup partition goes out of sync from the primary. 
so replication should not affect the running task on primary